### ISSUE TITLE
Data can get resurrected if flush and truncate run on the same time

### ISSUE THREAD

**Issue by haaawk** [scylladb/scylla#6555]

**Comment by haaawk** [comment]
@elcallio @gleb-cloudius could you please comment?

**Comment by slivne** [comment]
@elcallio ping ?

**Comment by elcallio** [comment]
Call me old fashioned, but I think that if one says there is a race, one should explain said race instead of insisting on a treasure hunt. 
Truncate flushes or clears memtables, picking up the RP at this point, then removes all sstables before or at said RP. Data can absolutely come in after this, but is there an actual race you have observed?

**Comment by haaawk** [comment]
Yes, there is a race w observed. This issue wasn't created after eyeballing the code but after running tests and observing this behavior. The test was as follows:
1. Have two nodes
2. Run write workload on them
3. Stop workload
4. Run truncate

This occasionally led to resurrected data but if we include nodetool flush between 3 and 4 then the resurrection never happened.

**Comment by slivne** [comment]
One theory is that hints were sent - so if we had a hints backlog and in parallel did a truncate (and hints were being replayed) then it may cause this.

whats the interaction between truncate and hints

**Comment by haaawk** [comment]
It's not about the hints @slivne. If the memtable is being flashed and truncate appears in the middle of that flash, the flash will continue after the truncate and data will be stored to disk despite truncate. At least, that's my understanding.

**Comment by elcallio** [comment]
Truncate only looks at whatever memtables says is its flush position, i.e. highest RP value. Then it prunes sstables. 
I can only assume this problem is because `cf.flush()/cf.clear()` does not wait for initiated/pending flush operations to finish. 
The code for flushing etc has been optimized and changed so many times without considering poor truncate...
I honestly have no idea how to ensure waiting for flushes in the code as it looks today.

**Comment by kostja** [comment]
@gleb should truncate go through raft log?

**Comment by raphaelsc** [comment]
> Truncate only looks at whatever memtables says is its flush position, i.e. highest RP value. Then it prunes sstables. I can only assume this problem is because `cf.flush()/cf.clear()` does not wait for initiated/pending flush operations to finish. The code for flushing etc has been optimized and changed so many times without considering poor truncate... I honestly have no idea how to ensure waiting for flushes in the code as it looks today.

the initiated/pending flush operation part was addressed by 3ad0f156b939de5ee507e122b03d3b24e2f4a98e

**Comment by gleb-cloudius** [comment]
> @gleb should truncate go through raft log?

Why?

**Comment by kostja** [comment]
That will generate an updated schema version, which we will hopefully validate when processing writes. So if the coordinator has got truncate in its raft log and replica hasn't, the replica will issue a read barrier, get and apply truncate, before accepting the write. This won't solve the problem of coordinated data resurrection in an entire replica set, if one imagines that truncate & flush race on all replicas, but at least all replicas will have the same view of when truncate happened with regard to writes they process.

**Comment by gleb-cloudius** [comment]
I do not see the point of discussing this here if it will not fix the problem described here.

**Comment by kostja** [comment]
Well, flush also could go through Raft log and then this issue will be solved...

**Comment by gleb-cloudius** [comment]
On Thu, Jan 20, 2022 at 07:35:30AM -0800, Konstantin Osipov wrote:
> Well, flush also could go through Raft log and then this issue will be solved...
> 
Flush is a local operation based on a memory pressure.

--
			Gleb.

**Comment by kostja** [comment]
Aren't we talking about nodetool flush here? I don't think local flush is of any issue because hardly local flush can correlate on all nodes.

**Comment by gleb-cloudius** [comment]
On Thu, Jan 20, 2022 at 10:34:22AM -0800, Konstantin Osipov wrote:
> Aren't we talking about nodetool flush here? I don't think local flush is of any issue because hardly local flush can correlate on all nodes. 
> 
No, we are taking about local flush. "nodetool flush" forces it and
solves the issue. Why do you think it had to correlate to cause the
issue?

Looks like the simple fix here is to flush on each node before executing
the truncate.

--
			Gleb.

**Comment by kostja** [comment]
https://github.com/scylladb/scylla/issues/6555#issuecomment-640446009 is using nodetool flush

**Comment by gleb-cloudius** [comment]
On Fri, Jan 21, 2022 at 01:08:48AM -0800, Konstantin Osipov wrote:
> https://github.com/scylladb/scylla/issues/6555#issuecomment-640446009 is using nodetool flush
> 

To solve the issue... The issue exists without it.

--
			Gleb.

**Comment by gleb-cloudius** [comment]
> Looks like the simple fix here is to flush on each node before executing the truncate.

And it looks like according to @elcallio there is no way to wait for the flush to complete, so the fix is not so simple. Anyway has nothing to do with Raft.
